### PR TITLE
Fix #666, broadcast no longer matches dims

### LINKED ISSUE #666: `CombineLayer` and others (`Data.get_common_data`) should use `broadcast_matches=False`
`CombineLayer` and others rely on `Data.get_common_data`.
 
`Data.get_common_data` uses `broadcast_matches=True` for the `DimensionTag.is_equal` opts. Such that you can match a tensor [B,1,T] with [B,4,T].

However, this relies on heuristics, and is not really needed. There is never a reason in RETURNN to create such broadcast dim [1] in the first place. So instead of [B,1,T] you would just have a tensor [B,T]. The broadcasting in RETURNN would automatically happen on non-existing dims. (`Data.copy_compatible_to` usually does that internally.) This would be much more consistent and reliable.

This needs a new behavior version (#508) because older configs rely on this.

### PR DESCRIPTION
**Pull request by Zettelkasten**
Needs #863 first (which should be ready now).

Fix #666.

Introduces a new behavior version (#508) because it would break some older configs.

### CODE REVIEW

**Comment by albertz** [comment]
Why do you say it needs #863 first but all tests are anyway passing without #863?

Anyway, I merged #863 now. So you should rebase here and resolve the conflict.

**Comment by Zettelkasten** [comment]
Ok, try now.
Sorry, I wasn't so clear about this, but the branch for this PR branched off #683, that is why the tests here also worked (and okay, we could've also closed #683 and just merged this directly).

File: tests/test_TFNetworkLayer.py [hunk 1]
```diff
@@ -3376,27 +3377,22 @@ def test_GatherLayer_constant_position():
           np.testing.assert_almost_equal(out_seqs[b, f1, f2], values_seqs[b, f1, position, f2])
 
 
-def test_GatherLayer_search_beam():
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:3379]: Why did you remove this test?

File: returnn/tf/util/data.py [hunk 2]
```diff
@@ -5039,16 +5039,18 @@ def get_batch_shape_dim_tags(self):
     return self.dim_tags
 
   @classmethod
-  def get_common_data(cls, sources, ignore_feature_dim=False, allow_broadcast_all_sources=NotSpecified):
+  def get_common_data(cls, sources, ignore_feature_dim=False, allow_broadcast_all_sources=NotSpecified, name=None):
```
Comment (albertz) [inline returnn/tf/util/data.py:5042]: Is there any specific reason this was added, or just to make it nicer?

File: tests/test_TFNetworkLayer.py [hunk 3]
```diff
@@ -920,7 +920,8 @@ def test_CombineLayer_broadcast():
     net_dict = {
       "lin1": {"class": "linear", "activation": "sigmoid", "n_out": 5, "from": "data:data"},
       "lin2": {"class": "linear", "activation": "sigmoid", "n_out": 1, "from": "data:data"},
-      "combine": {"class": "combine", "kind": "add", "from": ["lin1", "lin2"]},
+      "lin2_squeeze": {"class": "squeeze", "from": "lin2", "axis": "f"},
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:923]: I wonder whether we should somehow simplify this, and unify into a single layer. But not sure. Maybe not.

File: returnn/tf/layers/basic.py [hunk 4]
```diff
@@ -3771,7 +3774,7 @@ def get_out_data_from_opts(cls, name, axis, dim=1, sources=(), **kwargs):
       data = data.copy_as_batch_major()
     axis = cls._get_axis(data=data, axis=axis)
 
-    new_dim = SpatialDim("%s_expand_dims" % name, dim)
+    new_dim = (FeatureDim if init_axis.lower() == "f" else SpatialDim)("%s_expand_dims" % name, dim)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:3777]: I personally would have preferred to use `Dim(kind=... if ... else ..., description=..., dimension=dim)`. But maybe that's just personal preference.

**Comment by albertz** [comment]
Code style errors:
```
returnn/tf/layers/basic.py:5670: WARNING PyArgumentListInspection: Parameter 'name' unfilled
returnn/tf/layers/basic.py:6956: WEAK WARNING Pep8CodeStyle: E131: continuation line unaligned for hanging indent
returnn/tf/layers/basic.py:7207: WEAK WARNING Pep8CodeStyle: E131: continuation line unaligned for hanging indent
```

**Comment by albertz** [comment]
Isn't there some overlap with #866?

File: returnn/tf/util/data.py (hunk 2 above)
Comment (Zettelkasten) [reply returnn/tf/util/data.py:5042]: Ah yeah, my rebase messed this up. This is #866. I'll remove it from here again.

File: returnn/tf/layers/basic.py (hunk 4 above)
Comment (Zettelkasten) [reply returnn/tf/layers/basic.py]: I agree it's a bit nicer what you have, changed it.

File: tests/test_TFNetworkLayer.py (hunk 3 above)
Comment (Zettelkasten) [reply tests/test_TFNetworkLayer.py:923]: True. It is weird to think of `ExpandDimsLayer` as deprecated, but at the same time have valid use cases for `SqueezeLayer`.

Maybe via a special `out_dim` or so. But not too important.

**Review by albertz — APPROVED** [review]
